neo3/wallet/account.py

```diff
@@ -0,0 +1,258 @@
+from __future__ import annotations
+
+import base58
+import hashlib
+import unicodedata
+from typing import Optional
+
+from Crypto.Cipher import AES
+from neo3 import settings, contracts
+from neo3.core import types, to_script_hash
+from neo3.core.cryptography import ECPoint
+from neo3.core.cryptography import KeyPair
+
+# both constants below are used to encrypt/decrypt a private key to/from a nep2 key
+NEP_HEADER = bytes([0x01, 0x42])
+NEP_FLAG = bytes([0xe0])
+
+
+class Account:
+
+    def __init__(self, password: str,
+                 private_key: Optional[bytes] = None,
+                 watch_only: bool = False,
+                 address: Optional[str] = None
+                 ):
+        """
+        Instantiate an account. This constructor should only be directly called when it's desired to create a new
+        account using just a password and a randomly generated private key, otherwise, they should call another
+        function.
+        """
+
+        public_key: Optional[ECPoint] = None
+        encrypted_key: Optional[bytes] = None
+
+        if not watch_only:
+            key_pair: KeyPair
+
+            if private_key is None:
+                key_pair = KeyPair.generate()
+                private_key = key_pair.private_key
+            else:
+                key_pair = KeyPair(private_key)
+            encrypted_key = self.private_key_to_nep2(private_key, password)
+            script_hash = to_script_hash(contracts.Contract.create_signature_redeemscript(key_pair.public_key))
+            address = self.script_hash_to_address(script_hash)
+            public_key = key_pair.public_key
+
+        self.address = address
+        self.public_key = public_key
+        self.encrypted_key = encrypted_key
+
+    @property
+    def script_hash(self) -> types.UInt160:
+        return self.address_to_script_hash(self.address)
```

Comment:
Type checker fails because `address` is optional. While we know the correct use, somebody could ignore it and do something like 
```python
a = Account(password='', watch_only=True)
a.script_hash
```
We can solve this in the initialiser
```python
if watch_only and address is None:
   raise ValueError("Creating a watch only account requires an address")
```